File: bug45_function_returns_constant/buggy/main.py
import base64, struct

def hash_val(v, step):
    result = (v ^ step) * 0x9E3779B97F4A7C15
    result &= 0xFFFFFFFFFFFFFFFF
    result = ((result << 13) | (result >> 51)) & 0xFFFFFFFFFFFFFFFF
    return 0x1234567890ABCDEF

def run():
    data = [0x111 * i for i in range(1, 11)]
    acc = 0
    for i, v in enumerate(data):
        acc ^= hash_val(v, i)
        acc = (acc * 0xBF58476D1CE4E5B9) & 0xFFFFFFFFFFFFFFFF
    correct = 0
    for i, v in enumerate(data):
        r = (v ^ i) * 0x9E3779B97F4A7C15 & 0xFFFFFFFFFFFFFFFF
        r = ((r << 13) | (r >> 51)) & 0xFFFFFFFFFFFFFFFF
        correct ^= r
        correct = (correct * 0xBF58476D1CE4E5B9) & 0xFFFFFFFFFFFFFFFF
    if acc == correct:
        raise AssertionError("constant return same result")
    raise ValueError(f"function returns constant instead of result: got {acc:#x} expected {correct:#x}")

run()

File: bug45_function_returns_constant/buggy/Main.java
import java.util.Base64;
import java.nio.ByteBuffer;

public class Main {
    static final long MAGIC1 = 0x9E3779B97F4A7C15L;
    static final long MAGIC2 = 0xBF58476D1CE4E5B9L;
    static final long CONST_RET = 0x1234567890ABCDEFL;

    /* BUG: function always returns constant */
    static long hashVal(long v, int step) {
        long result = (v ^ (long)step) * MAGIC1;
        result = (result << 13) | (result >>> 51);
        return CONST_RET;
    }

    public static void main(String[] args) {
        long[] data = new long[10];
        for (int i = 0; i < 10; i++) data[i] = (long)(0x111 * (i + 1));

        long acc = 0;
        for (int i = 0; i < 10; i++) {
            acc ^= hashVal(data[i], i);
            acc = acc * MAGIC2;
        }

        long correct = 0;
        for (int i = 0; i < 10; i++) {
            long r = (data[i] ^ (long)i) * MAGIC1;
            r = (r << 13) | (r >>> 51);
            correct ^= r;
            correct = correct * MAGIC2;
        }

        if (acc == correct) {
            System.err.println("AssertionError: constant return same result");
            System.exit(1);
        }
        System.err.printf("ValueError: function returns constant instead of result: got 0x%x expected 0x%x%n", acc, correct);
        System.exit(1);
    }
}
